README.md

```diff
@@ -106,12 +110,62 @@ This term will reduce to:
 
 A match syntax for machine numbers.
 We match the case 0 and the case where the number is greater
-than 0, p binds the value of the matching number - 1:
+than 0, `n-1` binds the value of the matching number - 1:
+```rs
+Number.to_church = λn λf λx 
+  match n {
+    0: x
+    +: (f (Number.to_church n-1 f x))
+  }
+```
+
+It is possible to define Data types using `data`:
+```rs
+data Option = (Some val) | None
+```
+
+If the data type has a single possible value, it can be destructured using `let`:
+```rs
+data Boxed = (Box val)
+
+let (Box value) = boxed; value
+```
+
+Otherwise, there are two pattern syntaxes for matching on data types:
+
+```rs
+// Implicit bindings
+Option.map = λoption λf
+  match option {
+    Some: (Some (f option.val))
```

Comment:
Explain where the names of the implicit binds come from (matched variable . constructor name)